opentelemetry/src/global/trace.rs

```diff
@@ -285,9 +268,276 @@ where
+            sleep(Duration::from_secs(1));
+            let _ = sender2.send(format!("thread 2 :{:?}", tracer_provider()));
+        });
+
+        let first_resp = recv.recv().unwrap();
+        let second_resp = recv.recv().unwrap();
+        assert!(first_resp.contains("thread 2"));
+        assert!(second_resp.contains("thread 2"));
+    }
+
+    cfg_tokio! {
+        async fn test_set_provider_in_tokio() -> AssertWriter {
+            use crate::sdk::trace::TracerProvider;
+
+            let buffer = AssertWriter::new();
+
+            let exporter = crate::sdk::export::trace::stdout::Exporter::new(buffer.clone(), true);
+            let provider = TracerProvider::builder()
+                .with_exporter(exporter)
+                .build();
+
+            let _ = set_tracer_provider(provider);
+            let tracer = tracer("opentelemetery");
+
+            tracer.in_span("test", |_cx| {});
+
+            buffer
+        }
+    }
+
+    // When using `tokio::spawn` to spawn the worker task in batch processor
+    //
+    // multiple -> no shut down -> not export
+    // multiple -> shut down -> export
+    // single -> no shutdown -> not export
+    // single -> shutdown -> hang forever
+
+    // When using |fut| tokio::task::spawn_blocking(|| futures::executor::block_on(fut))
+    // to spawn the worker task in batch processor
+    //
+    // multiple -> no shutdown -> hang forever
+    // multiple -> shut down -> export
+    // single -> shut down -> export
+    // single -> no shutdown -> hang forever
+
+    // Test if the multiple thread tokio runtime could exit successfully when not force flushing spans
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    #[ignore]
+    #[cfg(feature = "tokio-support")]
+    async fn test_set_provider_multiple_thread_tokio() {
+        let assert_writer = test_set_provider_in_tokio().await;
+        assert_eq!(assert_writer.len(), 0);
+    }
+
+    // Test if the multiple thread tokio runtime could exit successfully when force flushing spans
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    #[ignore]
+    #[cfg(feature = "tokio-support")]
+    async fn test_set_provider_multiple_thread_tokio_shutdown() {
+        let assert_writer = test_set_provider_in_tokio().await;
+        shutdown_tracer_provider();
+        assert!(assert_writer.len() > 0);
+    }
+
+    // Test use simple processor in single thread tokio runtime.
+    // Expected to see the spans being exported to buffer
+    #[tokio::test]
+    #[ignore]
+    #[cfg(feature = "tokio-rt-current-thread")]
+    async fn test_set_provider_single_thread_tokio_with_simple_processor() {
+        use crate::sdk::trace::TracerProvider;
+
+        let assert_writer = AssertWriter::new();
+
+        let exporter =
+            crate::sdk::export::trace::stdout::Exporter::new(assert_writer.clone(), true);
+        let provider = TracerProvider::builder()
+            .with_simple_exporter(exporter)
```

Comment:
I like those tests. I found them mostly easy to read and understand what they test.

I wonder if we could move the tracer provider initialization here into a similar function as `test_set_provider_in_tokio`. It might seem unnecessary because we would only use it once. But it took me a bit to figure out why this test looks different than the others. The test name indicates that. But having a shorter function body might make it even easier.